### DISCUSSION TITLE
Create a generic constraint to represent self.

### DISCUSSION THREAD

**Discussion by tacosontitan** [dotnet/csharplang#5228] [Language Ideas]
```
public interface ITranslatable<T1, T2> where T1 : self {
    T2 Translate();
    T1 Translate(T2 input);
}
public class CentralDataModel { ... }
public sealed class A : ITranslatable<A, CentralDataModel> { ... }
public sealed class B : ITranslatable<B, CentralDataModel> { ... }
```

While inheritance or implementation of an interface is the *typical* route we should take here, there are edge cases (usually due to architecture requirements) that would benefit from such a constrant.

**Comment by 333fred** [comment]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-09-13.md#self-types is the most recent LDM where self types have come up. See also https://github.com/dotnet/csharplang/discussions/252 or https://github.com/dotnet/csharplang/discussions/2618.

**Reply by Joe4evr** [reply]
In the mean time, I have [an analyzer package](https://www.nuget.org/packages/JiiLib.Constraints/) that can emulate this for you (among a few other constraints not in the language at this time).
